# Data Scientist Associate Practical Exam Submission

Use this template to complete your analysis and write up your summary for submission.


## Task 1 
*Write your description here*
the column names and their respective answers are written:
# booking_id:
- a. The description provided in the table is "Nominal," but the actual description is "Continuous."
- b. There are no missing values in this column.
- c. Since the nature of the booking_id is typically a unique identifier, no action is needed to make values match the description.

# months_as_member:
- a. The description provided in the table is "Discrete," and the actual description is "Continuous," which matches.
- b. There are no missing values in this column.
- c. No action is needed since the actual description matches the expected description.
# 
# weight:
- a. The description provided in the table is "Continuous," and the actual description is "Continuous," which matches.
- b. There are 20 missing values in this column.
- c. To make values match the description, the missing values in the weight column have been replaced with the mean weight of the dataset.

# days_before:
- a. The description provided in the table is "Discrete," but the actual description is "Unknown."
- b. There are no missing values in this column.
- c. Since the actual description is "Unknown," no action is needed to make values match the description.

# day_of_week:
- a. The description provided in the table is "Nominal," but the actual description is "Unknown."
- b. There are no missing values in this column.
- c. To make values match the description, missing values in the day_of_week column have been replaced with "unknown."

# time:
- a. The description provided in the table is "Ordinal," but the actual description is "Unknown."
- b. There are no missing values in this column.
- c. To make values match the description, missing values in the time column have been replaced with "unknown."

# category:
- a. The description provided in the table is "Nominal," but the actual description is "Unknown."
- b. There are no missing values in this column.
- c. To make values match the description, missing values in the category column have been replaced with "unknown."

# attended:
- a. The description provided in the table is "Nominal," and the actual description is "Nominal," which matches.
- b. There are no missing values in this column.
- c. No action is needed since the actual description matches the expected description.


## Task 2
*Write your description here*
`import matplotlib.pyplot as plt
import seaborn as sns

# Create a countplot to visualize the number of bookings attended the class
plt.figure(figsize=(8, 6))
sns.countplot(data=df, x='attended')
plt.title('Attendance Distribution')
plt.xlabel('Attended')
plt.ylabel('Count')
plt.show()
`
![Screenshot 2023-10-06 120628](Screenshot%202023-10-06%20120628.png)
# task A

a. The category "0" of the variable "attended" has the most observations. This means that there are more bookings where members did not attend the class (category 0) compared to the bookings where members attended the class (category 1).
# task B
b. The observations are not balanced across categories of the variable "attended." Since category "0" has a significantly larger count compared to category "1," the dataset is imbalanced. There are more instances of members not attending the class than attending it. Imbalanced datasets can sometimes pose challenges when building predictive models, especially for rare events (in this case, attendance). It's important to keep this class imbalance in mind when working with the data and choosing an appropriate modeling approach.


## Task 3
*Write your description here*
![Screenshot 2023-10-06 121127](Screenshot%202023-10-06%20121127.png)

 it means that there is a longer tail on the right side of the distribution, and the majority of members have lower values of months as a member, with a few members having significantly higher values.

In practical terms, this suggests that most members have relatively short membership durations, while a smaller number of members have been with the fitness club for an extended period. Understanding the skewness of this distribution can be valuable for business decision-making, marketing strategies, and member retention efforts, as it highlights the diversity in membership duration.


## Task 4
`*Write your description here*
import seaborn as sns
import matplotlib.pyplot as plt

# Create a box plot to visualize the relationship between attendance and months_as_member
plt.figure(figsize=(10, 6))
sns.boxplot(data=df, x='attended', y='months_as_member')
plt.title('Relationship between Attendance and Months as a Member')
plt.xlabel('Attended')
plt.ylabel('Months as Member')
plt.xticks([0, 1], ['Not Attended', 'Attended'])

plt.show()`
From the box plot, we can observe the following relationship:

Attendance and Membership Duration: Among attendees, there are two distinct subgroups with different median membership durations: around 40 months and around 80 months. This suggests that attendees can be further divided based on their membership duration.
Comparison with Non-Attendees: On the other hand, non-attendees have a median membership duration of approximately 20 months, which is shorter than either subgroup of attendees. This indicates that non-attendees tend to be relatively newer members.
Outliers: Additionally, some attendees have significantly longer membership durations, reaching up to 120 months. These outliers may represent exceptional cases.
Overall, there appears to be a clear relationship between attendance and membership duration. Attendees generally have longer membership durations, and within the attendee group, there are distinct subgroups based on membership duration.

## Task 5
*Write your description here*
The type of machine learning problem for this task is a classification problem.

In a classification problem, the goal is to predict the category or class to which a data point (in this case, a member) belongs. In this specific scenario, the task is to predict whether a member will attend the fitness class or not. The target variable, "attended," is nominal and binary, taking values 0 (not attended) or 1 (attended). The objective is to build a predictive model that can classify members into one of these two categories based on various features such as "months_as_member," "weight," "days_before," "day_of_week," "time," and "category."

Classification is the appropriate type of problem for predicting a binary outcome like attendance (yes or no) in this context. The model will learn patterns and relationships in the dataset to make predictions about whether a member is likely to attend the class or not, based on the provided features.




## Task 6
*Write your description here*
The baseline model's performance metrics are as follows:

Accuracy: 0.7619 (approximately 76.19%)

Confusion Matrix:

True Negatives (TN): 189
False Positives (FP): 22
False Negatives (FN): 48
True Positives (TP): 35
Classification Report:

Precision for class 0: 0.80

Recall for class 0: 0.90

F1-score for class 0: 0.84

Precision for class 1: 0.61

Recall for class 1: 0.42

F1-score for class 1: 0.50

The accuracy of the model is 76.19%, which means it correctly predicts whether a member will attend or not in about 76.19% of the cases. The precision, recall, and F1-score for each class provide more detailed information about the model's performance, especially when dealing with imbalanced datasets like this one.

For class 0 (attended = 0):

Precision: 0.80
Recall: 0.90
F1-score: 0.84
For class 1 (attended = 1):

Precision: 0.61
Recall: 0.42
F1-score: 0.50
These metrics give insights into how well the model is performing for each class. In this case, the model has a higher precision and recall for class 0 (members who did not attend) compared to class 1 (members who attended). Further model tuning or exploring different algorithms may help improve the performance, especially for class 1, if necessary.

In [8]:
import pandas as pd

# Load the raw dataset
df = pd.read_csv("fitness_class_2212.csv")  

# Preprocess the data
# Handle missing values in the "weight" column by replacing "NA" with the overall average weight
average_weight = df["weight"].mean()
df["weight"].replace("NA", average_weight, inplace=True)

# Extract numerical values from the "days_before" column
df['days_before'] = df['days_before'].str.extract('(\d+)').astype(float)

# Perform one-hot encoding for the "day_of_week" column
df = pd.get_dummies(df, columns=['day_of_week'], drop_first=True)

# Map the "time" column to numerical values (e.g., "AM" -> 0, "PM" -> 1)
df["time"] = df["time"].map({"AM": 0, "PM": 1})

# Map the "category" column to numerical values (you can use one-hot encoding if needed)
category_mapping = {"Yoga": 0, "Aqua": 1, "Strength": 2, "HIIT": 3, "Cycling": 4}
df["category"] = df["category"].map(category_mapping)

# Convert the "attended" column to numeric values (0 and 1)
df["attended"] = df["attended"].replace({"-0": 0, 0: 0, 1: 1})

# Drop rows with missing values
df = df.dropna()

# Split the data into features (X) and target (y)
X = df.drop(columns=["attended"])
y = df["attended"]

# Now you can proceed with splitting the data and training your model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train a logistic regression model
baseline_model = LogisticRegression(random_state=42)
baseline_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = baseline_model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:")
print(conf_matrix)
print("Classification Report:")
print(class_report)


Accuracy: 0.7619047619047619
Confusion Matrix:
[[189  22]
 [ 48  35]]
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.90      0.84       211
           1       0.61      0.42      0.50        83

    accuracy                           0.76       294
   macro avg       0.71      0.66      0.67       294
weighted avg       0.75      0.76      0.75       294



## Task 7
*Write your description here*

In [9]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train a Random Forest Classifier
comparison_model = RandomForestClassifier(random_state=42)
comparison_model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = comparison_model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:")
print(conf_matrix)
print("Classification Report:")
print(class_report)



Accuracy: 0.7448979591836735
Confusion Matrix:
[[181  30]
 [ 45  38]]
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.86      0.83       211
           1       0.56      0.46      0.50        83

    accuracy                           0.74       294
   macro avg       0.68      0.66      0.67       294
weighted avg       0.73      0.74      0.74       294



## Task 8
*Write your description here*
Here's an explanation of why I chose these two models:

# Logistic Regression Model:

Logistic Regression is a simple yet effective classification algorithm that is well-suited for binary classification problems like predicting whether members will attend a fitness class (attended or not attended).
It provides probabilities for the target classes, making it easy to interpret the model's predictions.
Logistic Regression is computationally efficient and doesn't require a large amount of data to perform well.
It serves as a baseline model, and it's often a good starting point to establish a benchmark for more complex models.
# Random Forest Classifier:

Random Forest is an ensemble learning technique that combines multiple decision trees to improve predictive accuracy and reduce overfitting.
It can handle both binary and multiclass classification problems effectively.
Random Forests are robust to outliers, missing values, and irrelevant features, making them suitable for real-world datasets with noise and complexities.
They provide feature importance scores, which can help identify which features are most influential in making predictions.
Random Forests can capture complex relationships in the data and often outperform simpler models like Logistic Regression.
The choice between these two models is also influenced by the need for comparison. By demonstrating both Logistic Regression and Random Forest Classifier, I provide a comparison between a straightforward linear model and a more complex ensemble model. This comparison allows you to assess whether a more sophisticated model like Random Forest offers any significant improvement in prediction accuracy for this specific task.

## Task 9
*Write your description here*

In [10]:
# Start coding here... 
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score
from sklearn.model_selection import train_test_split

# Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize and train the Logistic Regression model
logistic_model = LogisticRegression(random_state=42)
logistic_model.fit(X_train, y_train)

# Make predictions using Logistic Regression
logistic_preds = logistic_model.predict(X_test)

# Initialize and train the Random Forest Classifier model
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

# Make predictions using Random Forest Classifier
rf_preds = rf_model.predict(X_test)

# Evaluate and compare the models
def evaluate_model(model_name, y_true, y_pred):
    accuracy = accuracy_score(y_true, y_pred)
    cm = confusion_matrix(y_true, y_pred)
    cr = classification_report(y_true, y_pred)
    roc_auc = roc_auc_score(y_true, y_pred)
    
    print(f"Performance Metrics for {model_name}:")
    print(f"Accuracy: {accuracy:.4f}")
    print("Confusion Matrix:")
    print(cm)
    print("Classification Report:")
    print(cr)
    print(f"ROC AUC Score: {roc_auc:.4f}")
    print("\n")

# Evaluate Logistic Regression
evaluate_model("Logistic Regression", y_test, logistic_preds)

# Evaluate Random Forest Classifier
evaluate_model("Random Forest Classifier", y_test, rf_preds)


Performance Metrics for Logistic Regression:
Accuracy: 0.7732
Confusion Matrix:
[[285  28]
 [ 72  56]]
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.91      0.85       313
           1       0.67      0.44      0.53       128

    accuracy                           0.77       441
   macro avg       0.73      0.67      0.69       441
weighted avg       0.76      0.77      0.76       441

ROC AUC Score: 0.6740


Performance Metrics for Random Forest Classifier:
Accuracy: 0.7574
Confusion Matrix:
[[275  38]
 [ 69  59]]
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.88      0.84       313
           1       0.61      0.46      0.52       128

    accuracy                           0.76       441
   macro avg       0.70      0.67      0.68       441
weighted avg       0.74      0.76      0.75       441

ROC AUC Score: 0.6698




## Task 10
*Write your description here*
the Logistic Regression model and the Random Forest Classifier model have fairly similar performance, but there are some differences to consider:

**Accuracy:** The Logistic Regression model has a slightly higher accuracy of approximately 0.7732, while the Random Forest Classifier has an accuracy of approximately 0.7574. In terms of overall correctness, the Logistic Regression model performs slightly better.

**Precision and Recall:** For both models, the precision and recall scores are relatively similar for class 0 (attended) and class 1 (did not attend). However, for class 1, the Logistic Regression model has slightly higher precision and recall compared to the Random Forest Classifier. This indicates that the Logistic Regression model is better at correctly identifying instances of class 1.

**F1-Score**: The F1-score balances precision and recall. For class 1, the Logistic Regression model has a higher F1-score (approximately 0.53) compared to the Random Forest Classifier (approximately 0.52). This suggests that the Logistic Regression model achieves a better balance between precision and recall for class 1.

**ROC AUC Score**: Both models have similar ROC AUC scores, with the Logistic Regression model having a score of approximately 0.6740, and the Random Forest Classifier having a score of approximately 0.6698. ROC AUC measures the ability of the model to distinguish between the two classes, and again, both models have similar performance in this aspect.

In summary, while both models perform relatively similarly in terms of accuracy and ROC AUC score, the Logistic Regression model performs slightly better in terms of precision, recall, and F1-score for class 1 (did not attend). Therefore, based on the provided metrics, the Logistic Regression model appears to be the better choice for predicting class attendance. However, it's important to consider other factors such as interpretability, computational complexity, and the specific goals of the application when choosing a model.

## ✅  When you have finished...
- Publish your Workspace using the option on the left
- Check the published version of your report:
	- Can you see everything you want us to grade?
    - Are all the graphics visible?
- Review the grading rubric. Have you included everything that will be graded?
- Head back to the [Certification dashboard](https://app.datacamp.com/certification) to submit your practical exam